### PR TITLE
Feature manage bookings UI #19

### LINKED ISSUE #19: FE: Manage Bookings UI
This view is for pet sitters who have requests coming in and have the ability to manage their bookings. 

Follow mockup. Refer to "Manage bookings" in the spec document

The mockup should include: 
-Calendar view
\- active upcoming bookings

### PR DESCRIPTION
**Pull request by MarianaES**
### What this PR does (required):
- Implement the bookings view with a calendar showing upcoming bookings as well as a summary of current and past bookings. 
- There is a settings button where the pet sitter can accept or decline each request.

### Screenshots / Videos (front-end only):
image

### Any information needed to test this feature (required):
- Install the following packages to make the DatePicker works.
```
@material-ui/pickers
@date-io/date-fns
```

### CODE REVIEW

File: server/controllers/request.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
+const Request = require("../models/Request");
+const asyncHandler = require("express-async-handler");
+
+// @route GET /request/all
+// @desc List of requests for logged in user
+// @access Private
```
Comment (ezeilo-su) [inline server/controllers/request.js:6]: Nice documentation

File: server/controllers/request.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    });
+  }
+
+  res.status(404);
+  throw new Error("No requests found");
+});
+
+// @route POST /request/create
+// @desc Create a new request
+// @access Private
```
Comment (ezeilo-su) [inline server/controllers/request.js:27]: Nice one too

File: server/controllers/request.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    res.status(201).json({ request });
+  }
+
+  res.status(400);
+  throw new Error("Error creating request");
+});
+
+// @route PUT /request/update/:id
+// @desc Update request dates, satus, serviceType and rating
+// @access Private
```
Comment (ezeilo-su) [inline server/controllers/request.js:55]: Nice documentation

File: server/controllers/request.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+      sitter: sitterId,
+      startDate,
+      endDate,
+      serviceType,
+      totalPrice,
+    });
+
+    await request.save();
+    res.status(201).json({ request });
+  }
```
Comment (ezeilo-su) [inline server/controllers/request.js:47]: :+1:

File: server/controllers/request.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+// @route PUT /request/update/:id
+// @desc Update request dates, satus, serviceType and rating
+// @access Private
+
+exports.updateRequest = asyncHandler(async (req, res, next) => {
+  const userId = req.user.id;
+  const requestId = req.params.id;
+  const input = req.body;
+
+  console.log({ userId, requestId, input });
```
Comment (ezeilo-su) [inline server/controllers/request.js:62]: Let's remove this comment

File: server/controllers/request.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+});
+
+// @route PUT /request/update/:id
+// @desc Update request dates, satus, serviceType and rating
+// @access Private
+
+exports.updateRequest = asyncHandler(async (req, res, next) => {
+  const userId = req.user.id;
+  const requestId = req.params.id;
+  const input = req.body;
```
Comment (ezeilo-su) [inline server/controllers/request.js:60]: Validate that the client supplied this

File: server/models/Request.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
+const mongoose = require("mongoose");
+
+const { Schema, model } = mongoose;
```
Comment (ezeilo-su) [inline server/models/Request.js:3]: Good destructuring :+1:

File: server/models/Request.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+      enum: ["BOARDING", "HOUSE_SITTING", "DAY_CARE", "WALKING"],
+    },
+    totalPrice: {
+      type: Number,
+    },
+    rating: {
+      type: Number,
+    },
+  },
+  { timestamps: true }
```
Comment (ezeilo-su) [inline server/models/Request.js:41]: :+1:  Nice use of mongoose `timestamps` property.

File: server/models/Request.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+);
+
+requestSchema.pre("validate", function (next) {
+  this.startDate > this.endDate
+    ? next(new Error("End date must be greater than start date"))
+    : next();
+});
```
Comment (ezeilo-su) [inline server/models/Request.js:48]: Good use of Mongoose pre hook

File: server/routes/request.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+const {
+  getRequests,
+  createRequest,
+  updateRequest,
+} = require("../controllers/request");
+
+router.route("/all").get(protect, getRequests);
+router.route("/create").post(protect, createRequest);
+router.route("/update/:id").put(protect, updateRequest);
```
Comment (ezeilo-su) [inline server/routes/request.js:13]: :+1:

**Review by sundayezeilo — COMMENTED** [review]
Nice work on this one, good-looking UI :+1:  Though you linked FE ticket but also added the BE here. Could you justify the decision? :smile:

**Comment by MarianaES** [comment]
> Nice work on this one, good-looking UI 👍 Though you linked FE ticket but also added the BE here. Could you justify the decision? 😄

**Comment by MarianaES** [comment]
> Nice work on this one, good-looking UI 👍 Though you linked FE ticket but also added the BE here. Could you justify the decision? 😄

It was a mistake, I was making the comparison to develop branch not the parent branch for this current branch. Now it should be showing only the FE files.

File: client/src/components/BookingItem/BookingItem.tsx [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+                </Menu>
+              </>
+            }
+            title={title.toUpperCase()}
+            disableTypography
+          />
+          <CardContent>
+            {/* TODO: Change harcoded date to booking(request) start date inside props. */}
+            <Typography variant="h5" className={classes.pos}>
+              5 April 2020, 10-12 AM
```
Comment (rajivtitus) [inline client/src/components/BookingItem/BookingItem.tsx:60]: In the future, as opposed to hardcoding data consider storing these values in an object and use the object's properties within this component :)

File: client/src/pages/Booking/Booking.tsx [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+  const classes = useStyles();
+
+  const { loggedInUser } = useAuth();
+  const history = useHistory();
+
+  const [bookedDays] = useState<number[]>([21, 25]);
+  const [selectedDate, setDate] = useState<MaterialUiPickersDate>(new Date());
+
+  if (loggedInUser === undefined) return <CircularProgress />;
+  if (!loggedInUser) {
```
Comment (rajivtitus) [inline client/src/pages/Booking/Booking.tsx:57]: Good thinking! We won't need this check though as we will eventually have a protected route which will take care of this for us :)

**Comment by rajivtitus** [comment]
@MarianaES Great work with this! Just some minor clean up comments (commented code), once that's done please tag me in 😄

File: client/src/components/BookingItem/BookingItem.tsx [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+import OngoingAndPastBookingItem from './OngoingAndPastBookingItem';
+
+interface Props {
+  upcoming?: boolean;
+}
+
+function BookingItem(props: Props): JSX.Element {
+  const { upcoming } = props;
+
+  return <>{upcoming ? <NextBookingItem /> : <OngoingAndPastBookingItem />}</>;
```
Comment (rajivtitus) [inline client/src/components/BookingItem/BookingItem.tsx:11]: Good job componentizing `NextBookingItem` and `OngoingAndPastBookingItem`! 
These components could've been conditionally rendered wherever applicable though - I don't think we need the `BookingItem` component - no action is needed here, just an FYI 😄

File: client/src/pages/Booking/Booking.tsx [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+  };
+
+  return (
+    <>
+      <Grid container spacing={6} direction="row" alignContent="center" alignItems="center" className={classes.root}>
+        <Grid item xs={12} sm={6}>
+          <Paper elevation={3}>
+            <Card>
+              <Typography variant="h6" className={classes.title}>
+                YOUR NEXT BOOKING:
```
Comment (rajivtitus) [inline client/src/pages/Booking/Booking.tsx]: Please avoid using CAPITAL CASE text as some screen readers interpret them as abbreviations. You can use a CSS class to transform text instead :)

File: client/src/pages/Booking/Booking.tsx [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+                renderDay={renderDay}
+              />
+            </MuiThemeProvider>
+          </MuiPickersUtilsProvider>
+        </Grid>
+        <Grid item xs={12} sm={6}>
+          <Paper elevation={3}>
+            <Card>
+              <Typography variant="h6" className={classes.title}>
+                CURRENT BOOKINGS:
```
Comment (rajivtitus) [inline client/src/pages/Booking/Booking.tsx]: Ditto as above and throughout!

**Review by rajivtitus — APPROVED** [review]
Great work here! 👍🏽 

Please address the pending comment prior to merging. Thanks :)

**Pull request merged**